### PR TITLE
Update to make sure preview mode works with getServerSideProps

### PR DESCRIPTION
**Pull request by ijjk**
This makes sure preview mode works correctly with `getServerSideProps` also and that `previewData` and `preview` are exposed in it's context. I also migrated the existing preview mode tests over for `getServerSideProps`

### CODE REVIEW

File: packages/next/next-server/server/next-server.ts [hunk 1]
```diff
@@ -905,6 +905,13 @@ export default class Server {
       })
     }
 
+    const previewData = tryGetPreviewData(
```
Comment (Timer) [inline packages/next/next-server/server/next-server.ts]: Let's skip this check unless the page is SSG or GSSP (as it parses cookies).

**Comment by ijjk** [comment]
# Stats from current PR

<details>
<summary><strong>Default Server Mode</strong></summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | zeit/next.js canary  | ijjk/next.js gssp/preview-mode | Change |
| - | - | - | - |
| buildDuration | 9.8s | 9.8s | -16ms |
| nodeModulesSize | 56.3 MB | 56.3 MB | ⚠️ +791 B |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | zeit/next.js canary  | ijjk/next.js gssp/preview-mode | Change |
| - | - | - | - |
| main-HASH.js gzip | 5.76 kB | 5.76 kB | ✓ |
| webpack-HASH.js gzip | 746 B | 746 B | ✓ |
| de003c3a9d30..67cb.js gzip | 9.69 kB | 9.69 kB | ✓ |
| framework.HASH.js gzip | 39.1 kB | 39.1 kB | ✓ |
| Overall change | 55.3 kB | 55.3 kB | ✓ |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons) Modern</strong></summary>

|  | zeit/next.js canary  | ijjk/next.js gssp/preview-mode | Change |
| - | - | - | - |
| main-HASH.module.js gzip | 4.79 kB | 4.79 kB | ✓ |
| webpack-HASH..dule.js gzip | 746 B | 746 B | ✓ |
| de003c3a9d30..dule.js gzip | 6.64 kB | 6.64 kB | ✓ |
| framework.HA..dule.js gzip | 39.1 kB | 39.1 kB | ✓ |
| Overall change | 51.3 kB | 51.3 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | zeit/next.js canary  | ijjk/next.js gssp/preview-mode | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 18.9 kB | 18.9 kB | ✓ |
| Overall change | 18.9 kB | 18.9 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | zeit/next.js canary  | ijjk/next.js gssp/preview-mode | Change |
| - | - | - | - |
| _app.js gzip | 1.07 kB | 1.07 kB | ✓ |
| _error.js gzip | 2.98 kB | 2.98 kB | ✓ |
| hooks.js gzip | 664 B | 664 B | ✓ |
| index.js gzip | 222 B | 222 B | ✓ |
| link.js gzip | 1.89 kB | 1.89 kB | ✓ |
| routerDirect.js gzip | 279 B | 279 B | ✓ |
| withRouter.js gzip | 278 B | 278 B | ✓ |
| Overall change | 7.38 kB | 7.38 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages Modern</strong></summary>

|  | zeit/next.js canary  | ijjk/next.js gssp/preview-mode | Change |
| - | - | - | - |
| _app.module.js gzip | 589 B | 589 B | ✓ |
| _error.module.js gzip | 2.08 kB | 2.08 kB | ✓ |
| hooks.module.js gzip | 370 B | 370 B | ✓ |
| index.module.js gzip | 212 B | 212 B | ✓ |
| link.module.js gzip | 1.48 kB | 1.48 kB | ✓ |
| routerDirect..dule.js gzip | 271 B | 271 B | ✓ |
| withRouter.m..dule.js gzip | 270 B | 270 B | ✓ |
| Overall change | 5.27 kB | 5.27 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | zeit/next.js canary  | ijjk/next.js gssp/preview-mode | Change |
| - | - | - | - |
| _buildManifest.js gzip | 61 B | 61 B | ✓ |
| _buildManife..dule.js gzip | 61 B | 61 B | ✓ |
| Overall change | 122 B | 122 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | zeit/next.js canary  | ijjk/next.js gssp/preview-mode | Change |
| - | - | - | - |
| index.html gzip | 936 B | 936 B | ✓ |
| link.html gzip | 946 B | 946 B | ✓ |
| withRouter.html gzip | 934 B | 934 B | ✓ |
| Overall change | 2.82 kB | 2.82 kB | ✓ |

</details>

</details>
<hr/>
<details>
<summary><strong>Serverless Mode</strong> (Decrease detected ✓)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | zeit/next.js canary  | ijjk/next.js gssp/preview-mode | Change |
| - | - | - | - |
| buildDuration | 10.3s | 9.8s | -477ms |
| nodeModulesSize | 56.3 MB | 56.3 MB | ⚠️ +791 B |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | zeit/next.js canary  | ijjk/next.js gssp/preview-mode | Change |
| - | - | - | - |
| main-HASH.js gzip | 5.76 kB | 5.76 kB | ✓ |
| webpack-HASH.js gzip | 746 B | 746 B | ✓ |
| de003c3a9d30..67cb.js gzip | 9.69 kB | 9.69 kB | ✓ |
| framework.HASH.js gzip | 39.1 kB | 39.1 kB | ✓ |
| Overall change | 55.3 kB | 55.3 kB | ✓ |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons) Modern</strong></summary>

|  | zeit/next.js canary  | ijjk/next.js gssp/preview-mode | Change |
| - | - | - | - |
| main-HASH.module.js gzip | 4.79 kB | 4.79 kB | ✓ |
| webpack-HASH..dule.js gzip | 746 B | 746 B | ✓ |
| de003c3a9d30..dule.js gzip | 6.64 kB | 6.64 kB | ✓ |
| framework.HA..dule.js gzip | 39.1 kB | 39.1 kB | ✓ |
| Overall change | 51.3 kB | 51.3 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | zeit/next.js canary  | ijjk/next.js gssp/preview-mode | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 18.9 kB | 18.9 kB | ✓ |
| Overall change | 18.9 kB | 18.9 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | zeit/next.js canary  | ijjk/next.js gssp/preview-mode | Change |
| - | - | - | - |
| _app.js gzip | 1.07 kB | 1.07 kB | ✓ |
| _error.js gzip | 2.98 kB | 2.98 kB | ✓ |
| hooks.js gzip | 664 B | 664 B | ✓ |
| index.js gzip | 222 B | 222 B | ✓ |
| link.js gzip | 1.89 kB | 1.89 kB | ✓ |
| routerDirect.js gzip | 279 B | 279 B | ✓ |
| withRouter.js gzip | 278 B | 278 B | ✓ |
| Overall change | 7.38 kB | 7.38 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages Modern</strong></summary>

|  | zeit/next.js canary  | ijjk/next.js gssp/preview-mode | Change |
| - | - | - | - |
| _app.module.js gzip | 589 B | 589 B | ✓ |
| _error.module.js gzip | 2.08 kB | 2.08 kB | ✓ |
| hooks.module.js gzip | 370 B | 370 B | ✓ |
| index.module.js gzip | 212 B | 212 B | ✓ |
| link.module.js gzip | 1.48 kB | 1.48 kB | ✓ |
| routerDirect..dule.js gzip | 271 B | 271 B | ✓ |
| withRouter.m..dule.js gzip | 270 B | 270 B | ✓ |
| Overall change | 5.27 kB | 5.27 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | zeit/next.js canary  | ijjk/next.js gssp/preview-mode | Change |
| - | - | - | - |
| _buildManifest.js gzip | 61 B | 61 B | ✓ |
| _buildManife..dule.js gzip | 61 B | 61 B | ✓ |
| Overall change | 122 B | 122 B | ✓ |

</details>

<details>
<summary><strong>Serverless bundles</strong> Overall decrease ✓</summary>

|  | zeit/next.js canary  | ijjk/next.js gssp/preview-mode | Change |
| - | - | - | - |
| _error.js gzip | 292 kB | 292 kB | ⚠️ +156 B |
| 404.html gzip | 1.34 kB | 1.34 kB | ✓ |
| hooks.html gzip | 975 B | 975 B | ✓ |
| index.js gzip | 292 kB | 292 kB | -266 B |
| link.js gzip | 299 kB | 299 kB | ⚠️ +184 B |
| routerDirect.js gzip | 299 kB | 298 kB | -1.03 kB |
| withRouter.js gzip | 298 kB | 298 kB | ⚠️ +1 B |
| Overall change | 1.48 MB | 1.48 MB | -959 B |

</details>

</details>

**Comment by ijjk** [comment]
## Failing test suites 
- test/integration/custom-server/test/index.test.js

File: packages/next/next-server/server/render.tsx [hunk 2]
```diff
@@ -465,12 +465,12 @@ export async function renderToHTML(
       router,
       ctx,
     })
+    // Reads of this are cached on the `req` object, so this should resolve
+    // instantly. There's no need to pass this data down from a previous
+    // invoke, where we'd have to consider server & serverless.
+    const previewData = tryGetPreviewData(req, res, previewProps)
```
Comment (Timer) [inline packages/next/next-server/server/render.tsx]: Ditto:
> Let's skip this check unless the page is SSG or GSSP (as it parses cookies).

**Review by Timer — CHANGES_REQUESTED** [review]

File: packages/next/next-server/server/next-server.ts (hunk 1 above)
Comment (ijjk) [reply packages/next/next-server/server/next-server.ts]: Updated

File: packages/next/next-server/server/render.tsx (hunk 2 above)
Comment (ijjk) [reply packages/next/next-server/server/render.tsx]: Updated

**Comment by ijjk** [comment]
# Stats from current PR

<details>
<summary><strong>Default Server Mode</strong></summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | zeit/next.js canary  | ijjk/next.js gssp/preview-mode | Change |
| - | - | - | - |
| buildDuration | 9.4s | 9.6s | ⚠️ +129ms |
| nodeModulesSize | 56.3 MB | 56.3 MB | ⚠️ +1.29 kB |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | zeit/next.js canary  | ijjk/next.js gssp/preview-mode | Change |
| - | - | - | - |
| main-HASH.js gzip | 5.76 kB | 5.76 kB | ✓ |
| webpack-HASH.js gzip | 746 B | 746 B | ✓ |
| de003c3a9d30..67cb.js gzip | 9.69 kB | 9.69 kB | ✓ |
| framework.HASH.js gzip | 39.1 kB | 39.1 kB | ✓ |
| Overall change | 55.3 kB | 55.3 kB | ✓ |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons) Modern</strong></summary>

|  | zeit/next.js canary  | ijjk/next.js gssp/preview-mode | Change |
| - | - | - | - |
| main-HASH.module.js gzip | 4.79 kB | 4.79 kB | ✓ |
| webpack-HASH..dule.js gzip | 746 B | 746 B | ✓ |
| de003c3a9d30..dule.js gzip | 6.64 kB | 6.64 kB | ✓ |
| framework.HA..dule.js gzip | 39.1 kB | 39.1 kB | ✓ |
| Overall change | 51.3 kB | 51.3 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | zeit/next.js canary  | ijjk/next.js gssp/preview-mode | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 18.9 kB | 18.9 kB | ✓ |
| Overall change | 18.9 kB | 18.9 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | zeit/next.js canary  | ijjk/next.js gssp/preview-mode | Change |
| - | - | - | - |
| _app.js gzip | 1.07 kB | 1.07 kB | ✓ |
| _error.js gzip | 2.98 kB | 2.98 kB | ✓ |
| hooks.js gzip | 664 B | 664 B | ✓ |
| index.js gzip | 222 B | 222 B | ✓ |
| link.js gzip | 1.89 kB | 1.89 kB | ✓ |
| routerDirect.js gzip | 279 B | 279 B | ✓ |
| withRouter.js gzip | 278 B | 278 B | ✓ |
| Overall change | 7.38 kB | 7.38 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages Modern</strong></summary>

|  | zeit/next.js canary  | ijjk/next.js gssp/preview-mode | Change |
| - | - | - | - |
| _app.module.js gzip | 589 B | 589 B | ✓ |
| _error.module.js gzip | 2.08 kB | 2.08 kB | ✓ |
| hooks.module.js gzip | 370 B | 370 B | ✓ |
| index.module.js gzip | 212 B | 212 B | ✓ |
| link.module.js gzip | 1.48 kB | 1.48 kB | ✓ |
| routerDirect..dule.js gzip | 271 B | 271 B | ✓ |
| withRouter.m..dule.js gzip | 270 B | 270 B | ✓ |
| Overall change | 5.27 kB | 5.27 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | zeit/next.js canary  | ijjk/next.js gssp/preview-mode | Change |
| - | - | - | - |
| _buildManifest.js gzip | 61 B | 61 B | ✓ |
| _buildManife..dule.js gzip | 61 B | 61 B | ✓ |
| Overall change | 122 B | 122 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | zeit/next.js canary  | ijjk/next.js gssp/preview-mode | Change |
| - | - | - | - |
| index.html gzip | 936 B | 936 B | ✓ |
| link.html gzip | 946 B | 946 B | ✓ |
| withRouter.html gzip | 934 B | 934 B | ✓ |
| Overall change | 2.82 kB | 2.82 kB | ✓ |

</details>

</details>
<hr/>
<details>
<summary><strong>Serverless Mode</strong> (Decrease detected ✓)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | zeit/next.js canary  | ijjk/next.js gssp/preview-mode | Change |
| - | - | - | - |
| buildDuration | 10.4s | 10.2s | -245ms |
| nodeModulesSize | 56.3 MB | 56.3 MB | ⚠️ +1.29 kB |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | zeit/next.js canary  | ijjk/next.js gssp/preview-mode | Change |
| - | - | - | - |
| main-HASH.js gzip | 5.76 kB | 5.76 kB | ✓ |
| webpack-HASH.js gzip | 746 B | 746 B | ✓ |
| de003c3a9d30..67cb.js gzip | 9.69 kB | 9.69 kB | ✓ |
| framework.HASH.js gzip | 39.1 kB | 39.1 kB | ✓ |
| Overall change | 55.3 kB | 55.3 kB | ✓ |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons) Modern</strong></summary>

|  | zeit/next.js canary  | ijjk/next.js gssp/preview-mode | Change |
| - | - | - | - |
| main-HASH.module.js gzip | 4.79 kB | 4.79 kB | ✓ |
| webpack-HASH..dule.js gzip | 746 B | 746 B | ✓ |
| de003c3a9d30..dule.js gzip | 6.64 kB | 6.64 kB | ✓ |
| framework.HA..dule.js gzip | 39.1 kB | 39.1 kB | ✓ |
| Overall change | 51.3 kB | 51.3 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | zeit/next.js canary  | ijjk/next.js gssp/preview-mode | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 18.9 kB | 18.9 kB | ✓ |
| Overall change | 18.9 kB | 18.9 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | zeit/next.js canary  | ijjk/next.js gssp/preview-mode | Change |
| - | - | - | - |
| _app.js gzip | 1.07 kB | 1.07 kB | ✓ |
| _error.js gzip | 2.98 kB | 2.98 kB | ✓ |
| hooks.js gzip | 664 B | 664 B | ✓ |
| index.js gzip | 222 B | 222 B | ✓ |
| link.js gzip | 1.89 kB | 1.89 kB | ✓ |
| routerDirect.js gzip | 279 B | 279 B | ✓ |
| withRouter.js gzip | 278 B | 278 B | ✓ |
| Overall change | 7.38 kB | 7.38 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages Modern</strong></summary>

|  | zeit/next.js canary  | ijjk/next.js gssp/preview-mode | Change |
| - | - | - | - |
| _app.module.js gzip | 589 B | 589 B | ✓ |
| _error.module.js gzip | 2.08 kB | 2.08 kB | ✓ |
| hooks.module.js gzip | 370 B | 370 B | ✓ |
| index.module.js gzip | 212 B | 212 B | ✓ |
| link.module.js gzip | 1.48 kB | 1.48 kB | ✓ |
| routerDirect..dule.js gzip | 271 B | 271 B | ✓ |
| withRouter.m..dule.js gzip | 270 B | 270 B | ✓ |
| Overall change | 5.27 kB | 5.27 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | zeit/next.js canary  | ijjk/next.js gssp/preview-mode | Change |
| - | - | - | - |
| _buildManifest.js gzip | 61 B | 61 B | ✓ |
| _buildManife..dule.js gzip | 61 B | 61 B | ✓ |
| Overall change | 122 B | 122 B | ✓ |

</details>

<details>
<summary><strong>Serverless bundles</strong> Overall decrease ✓</summary>

|  | zeit/next.js canary  | ijjk/next.js gssp/preview-mode | Change |
| - | - | - | - |
| _error.js gzip | 292 kB | 292 kB | -196 B |
| 404.html gzip | 1.34 kB | 1.34 kB | ✓ |
| hooks.html gzip | 975 B | 975 B | ✓ |
| index.js gzip | 293 kB | 292 kB | -186 B |
| link.js gzip | 299 kB | 299 kB | -21 B |
| routerDirect.js gzip | 299 kB | 299 kB | ⚠️ +83 B |
| withRouter.js gzip | 299 kB | 298 kB | -1.24 kB |
| Overall change | 1.48 MB | 1.48 MB | -1.56 kB |

</details>

</details>

**Review by timneutkens — APPROVED** [review]

**Pull request merged**